crates/math/src/arith_expr.rs

```diff
@@ -727,116 +229,367 @@ impl<F: Field> ArithCircuit<F> {
+	}
+
+	pub fn optimize(mut self) -> Self {
+		self.optimize_constants();
+		self.deduplicate_steps();
+		self.compress_unused_steps();
+
+		self
 	}
 }
 
-impl<F: Field, FTgt: Field + From<F>> From<&ArithCircuit<F>> for ArithExpr<FTgt> {
-	fn from(circuit: &ArithCircuit<F>) -> Self {
-		fn visit_step<F: Field, FTgt: Field + From<F>>(
+impl<F: Field> Display for ArithExpr<F> {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+		fn display_step<F: Field>(
 			step: usize,
-			cached_exrs: &mut [Option<Arc<ArithExpr<FTgt>>>],
 			steps: &[ArithCircuitStep<F>],
-		) -> Arc<ArithExpr<FTgt>> {
-			if let Some(node) = cached_exrs[step].as_ref() {
-				return Arc::clone(node);
-			}
-
-			let node = match &steps[step] {
-				ArithCircuitStep::Const(value) => Arc::new(ArithExpr::Const((*value).into())),
-				ArithCircuitStep::Var(index) => Arc::new(ArithExpr::Var(*index)),
+			f: &mut fmt::Formatter<'_>,
+		) -> Result<(), fmt::Error> {
+			match &steps[step] {
+				ArithCircuitStep::Const(value) => write!(f, "{}", value),
+				ArithCircuitStep::Var(index) => write!(f, "x{}", index),
 				ArithCircuitStep::Add(left, right) => {
-					let left = visit_step(*left, cached_exrs, steps);
-					let right = visit_step(*right, cached_exrs, steps);
-					Arc::new(ArithExpr::Add(left, right))
+					write!(f, "(")?;
+					display_step(*left, steps, f)?;
+					write!(f, " + ")?;
+					display_step(*right, steps, f)?;
+					write!(f, ")")
 				}
 				ArithCircuitStep::Mul(left, right) => {
-					let left = visit_step(*left, cached_exrs, steps);
-					let right = visit_step(*right, cached_exrs, steps);
-					Arc::new(ArithExpr::Mul(left, right))
+					write!(f, "(")?;
+					display_step(*left, steps, f)?;
+					write!(f, " * ")?;
+					display_step(*right, steps, f)?;
+					write!(f, ")")
 				}
 				ArithCircuitStep::Pow(base, exp) => {
-					let base = visit_step(*base, cached_exrs, steps);
-					Arc::new(ArithExpr::Pow(base, *exp))
+					write!(f, "(")?;
+					display_step(*base, steps, f)?;
+					write!(f, ")^{}", exp)
 				}
-			};
+			}
+		}
 
-			cached_exrs[step] = Some(Arc::clone(&node));
-			node
+		display_step(self.steps.len() - 1, &self.steps, f)
+	}
+}
+
+impl<F: Field> PartialEq for ArithExpr<F> {
+	fn eq(&self, other: &Self) -> bool {
+		StepNode {
+			index: self.steps.len() - 1,
+			steps: &self.steps,
+		} == StepNode {
+			index: other.steps.len() - 1,
+			steps: &other.steps,
 		}
+	}
+}
 
-		let mut cached_exprs = vec![None; circuit.steps.len()];
-		let root = visit_step(circuit.steps.len() - 1, &mut cached_exprs, &circuit.steps);
+impl<F: Field> Add for ArithExpr<F> {
+	type Output = Self;
 
-		// to remove the extra copy of the root node
-		drop(cached_exprs);
+	fn add(mut self, rhs: Self) -> Self {
+		self += rhs;
+		self
+	}
+}
+
+impl<F: Field> AddAssign for ArithExpr<F> {
+	fn add_assign(&mut self, mut rhs: Self) {
+		let old_len = self.steps.len();
+		add_offset(&mut rhs.steps, old_len);
```

Comment:
Won't this duplicate nodes if the lhs and rhs are expressions sharing a common subexpression?

Reply:
Yes it will. The workflow is to build the expression and call `optimize` once at the end (it is called at the moment creation of the ArithCircuitPoly). This will extract all the duplicate nodes and compress the circuit

Reply:
This looks like a regression in functionality to me. Please keep ArithExpr, which dedups by construction.

Reply:
We can automatically call the deduplication after each operation which would give the similar result.

But TBH calling `optimize` before we're constructing some computation object for the expression (`ArithCircuitPoly` or generating instruction for FPGA) is necessary anyway. Because after we've constructed the polynomials manually we can apply arithmetic operations to them in compositions, partially substitute variables with constants, etc. So the best place to make the constant folding/extract common steps is there.

Or maybe I'm missing something and there are some situations when we want to ensure that certain steps will be shared before that?